### PR TITLE
[mdk] Add a starter Github action workflow file

### PR DESCRIPTION
**Pull request by williewillus**
[mdk] Add a starter Github action workflow file

Just a small template that helps modders that choose to open-source their mods vet their
and others' changes.  All it does is validate that no one tampered with the gradle wrapper
(this is a possible attack vector that has been exploited before, iirc), and that the mod
compiles.

I think it'd be good to make gametests run by default too, but build.gradle says that
having no tests crashes the run config, so I didn't add one yet.

### CODE REVIEW

**Comment by williewillus** [comment]
Because I anticipate the bikeshedding crowd already:

* "Why Github Actions????"
Because, for better or worse, GitHub is where most FOSS activity goes and the code forge most people use. I don't like GitHub, I don't like how dominant it is, but providing a paved path for beginners is more important than my personal beliefs. This builtin action is not going to cost any money and provides good value for amateur modders. Public repo GH Action minutes are more than enough (All of Violet Moon's mods use the free actions and we've never had to pay). And those with private repos should be smart enough to figure out how to delete the actions file.
* "How about \<insert other code forge or CI system here\>??"
Feel free to open your own PR.

**Comment by marchermans** [comment]
No.

First and foremost, on a personal opinion, I don't like GitHub Actions as a CI environment.

However that said, this enables the CI environment (given that you literally put it in the right directory for it to trigger) immediatly, and that is not something we are willing to support.

Please put this in a directory where the user has to make a choice to enable it (AKA not in .github).

Second, given that many people use many different CI systems, I would think it would be a better idea to include different examples, not just github.
Could you also include one for jenkins and teamcity?

File: mdk/.github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+          ~/.gradle/caches
+          ~/.gradle/wrapper
+        key: ${{ runner.os }}-gradle-${{ hashFiles('build.gradle', '**/gradle-wrapper.properties') }}
+    - name: Build with Gradle
+      run: ./gradlew build
+    - name: Archive Forge Binaries
+      uses: actions/upload-artifact@v3
+      with:
+        name: Forge Jars
+        path: build/libs/*.jar
```
Comment (marchermans) [inline mdk/.github/workflows/ci.yml:30]: Move this file into a template directory.
And add examples for other CI environments.

**Review by marchermans — CHANGES_REQUESTED** [review]

**Comment by williewillus** [comment]
> First and foremost, on a personal opinion, I don't like GitHub Actions as a CI environment.

Great. I don't like it very much either. The point here is reach, not what we both personally like more.

> However that said, this enables the CI environment (given that you literally put it in the right directory for it to trigger) immediatly, and that is not something we are willing to support.

The point is to have a paved, sensible-by-default thing that your typical high-school-modder can gain value from. Not enabling it default is no better than having nothing.

>  Could you also include one for jenkins and teamcity?

I have no knowledge of those two CI systems, that's why I mentioned in my comment that other contributions are welcome for other CI systems. Let's keep the PR small and focused on one thing at a time.

**Comment by marchermans** [comment]
I talked this through internally.
To get this merged, Forge has some requirements that need to be met:
1) The build setup needs to be significantly improved, simply building and calling it a day is not enough for this to get merged into the MDK. Make a template repo or an action if this is all the time you want to invest.
  a) The build needs to support test reporting
  b) The build needs to require signing jars with JarSigner!
  c) The build needs to be aware of the environment it is running in, and disable certain steps if it is running for a PR for example
  d) The build needs to release the artifacts to the three major platforms currently involved:
    - CurseForge
    - ModRinth
    - GitHub
  

2) Since Forge does not take sides or recommends a CI environment to use, and never will, this needs to be an OPT-IN if the build does not support JarSigner.
3) A PR needs to be opened to the documentation repo that indicates how to set this build up and how to use it. What to pay attention to and how to handle PR builds and their security with regards to the use of secrets.

**Comment by williewillus** [comment]
re 1. That's multiple months of work I'm not willing to sign up for, especially since it will involve dealing with lots of bikeshedding and potentially with hostile personalities. You're asking me to boil the ocean when I just want to push things a small step forward. Small incremental steps forward is how real progress is made.
Not to mention the reservations I have about jarsigner in #9598.

re 2. An important part of advancing the status quo is making things safe and easy to use by default. If it's not enabled by default, it might as well not exist to the people I'm trying to help with this change. Also, it is somewhat of a misdirective point to say "Forge doesn't want to recommend a CI provider", when Forge is hosted on GitHub. Why don't we mirror the repo to other code forges too, to avoid "recommending a provider"?

So never mind. If a template repository is created I'll retarget this there, but until that happens i'll just let this sit here.

**Pull request closed without merging**